rllib/env/multi_agent_episode.py

```diff
@@ -755,6 +753,108 @@ def finalize(
 
         return self
 
+    def concat_episode(self, other: "MultiAgentEpisode") -> None:
+        """Adds the given `other` MultiAgentEpisode to the right side of self.
+
+        In order for this to work, both chunks (`self` and `other`) must fit
+        together. This is checked by the IDs (must be identical), the time step counters
+        (`self.t` must be the same as `episode_chunk.t_started`), as well as the
+        observations/infos at the concatenation boundaries (`self.observations[-1]`
+        must match `episode_chunk.observations[0]`). Also, `self.is_done` must
+        not be True, meaning `self.is_terminated` and `self.is_truncated` are both
+        False.
+
+        Args:
+            other: The other `MultiAgentEpisode` to be concatenated to this one.
+
+        Returns: A `MultiAgentEpisode` instance containing the concatenated data
+            from both episodes (`self` and `other`).
+        """
+        # Make sure the IDs match.
+        assert other.id_ == self.id_
+        # NOTE (sven): This is what we agreed on. As the replay buffers must be
+        # able to concatenate.
+        assert not self.is_done
+        # Make sure the timesteps match.
+        assert self.env_t == other.env_t_started
+        # Validate `other`.
+        other.validate()
+
+        # Concatenate the individual SingleAgentEpisodes from both chunks.
+        all_agent_ids = set(self.agent_ids) | set(other.agent_ids)
+        for agent_id in all_agent_ids:
+            sa_episode = self.agent_episodes.get(agent_id)
+
+            # If agent is only in the new episode chunk.
+            if sa_episode is None:
+                # Then store all agent data from the new episode chunk in self.
+                self.agent_episodes[agent_id] = other.agent_episodes[agent_id]
+                # Do not forget the env to agent timestep mapping.
+                self.env_t_to_agent_t[agent_id] = other.env_t_to_agent_t[agent_id]
+                # Add the agent's starting timestep.
+                self.agent_t_started[agent_id] = other.agent_t_started[agent_id]
+
+            # If the agent was done in the first episode chunk, continue.
+            elif sa_episode.is_done:
+                continue
+
+            # If the agent has data in both chunks, simply concatenate on the
+            # single-agent level.
+            elif agent_id in other.agent_episodes:
+                # If `self` has hanging agent values -> Add these to `other`'s agent
+                # SingleAgentEpisode (as a new timestep) and only then concatenate.
+                # Otherwise, the concatentaion would fail b/c of missing data.
+                if agent_id in self._agent_buffered_actions:
+                    assert agent_id in self._agent_buffered_extra_model_outputs
+                    sa_episode.add_env_step(
+                        observation=other.agent_episodes[agent_id].get_observations(0),
```

Comment:
Is `indices=0` the first "new" one from the `other` episode chunk for agents that were missing the next observation? The last observation in `self` is then in the `other.observations`' lookback buffer, correct?

Reply:
Nope, observations (and infos) always overlap by one ts (regardless of any lookback, so even if lookback=0)!

So for a multi agent episode, if you do e.g. `self.cut()`, the returned successor's first observation (multi-agent dict) is identical to `self`'s last observation (multi-agent dict). Same principal as for single-agent episodes.

Reply:
However, here, we have to concatenate single-agent episodes (that sit inside a multi-agent one). For those single-agent episodes, in case agents are not always stepping with each env(!) step, they might "miss" this overlap. In this case, just for the purpose of concatenating the individual single-agent episodes, we have to "fix" this and add the overlap-timestep to the first SA episodes (so its observations overlap with the second SA episode' by 1 ts).